component/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java

```diff
@@ -0,0 +1,170 @@
+                                                 AuthenticationContext authenticationContext)
+            throws AuthenticationFailedException {
+        String authEndpoint = null;
+        try {
+            authEndpoint = X509CertificateConstants.AUTH_ENDPOINT;
+            String queryParams = FrameworkUtils.getQueryStringWithFrameworkContextId(authenticationContext
+                    .getQueryParams(), authenticationContext.getCallerSessionKey(), authenticationContext
+                    .getContextIdentifier());
+            httpServletResponse.sendRedirect(httpServletResponse.encodeRedirectURL(authEndpoint + ("?" + queryParams)));
+            if (log.isDebugEnabled()) {
+                log.debug("Request send to " + authEndpoint);
+            }
+        } catch (IOException e) {
+            throw new AuthenticationFailedException("Error when sending to the login page :" + authEndpoint, e);
+        }
+    }
+
+    /**
+     * Validate the certificate.
+     *
+     * @param httpServletRequest    http request
+     * @param httpServletResponse   http response
+     * @param authenticationContext authentication context
+     * @throws AuthenticationFailedException
+     */
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest httpServletRequest,
+                                                 HttpServletResponse httpServletResponse,
+                                                 AuthenticationContext authenticationContext)
+            throws AuthenticationFailedException {
+        AuthenticatedUser authenticatedUser = getUsername(authenticationContext);
+        String userName = authenticatedUser.getAuthenticatedSubjectIdentifier();
+        Object object = httpServletRequest.getAttribute(X509CertificateConstants.X_509_CERTIFICATE);
+        if (object != null) {
+            X509Certificate[] certificates = (X509Certificate[]) object;
+            if (certificates.length > 0) {
+                if (log.isDebugEnabled()) {
+                    log.debug("X509 Certificate Checking in servlet is done! ");
+                }
+                X509Certificate certs[] = (X509Certificate[]) object;
+                X509Certificate cert = certs[0];
+                byte[] data;
+                try {
+                    data = cert.getEncoded();
+                } catch (CertificateEncodingException e) {
+                    throw new AuthenticationFailedException("Encoded certificate not found", e);
+                }
+                X509CertificateUtil certificateUtil = new X509CertificateUtil();
+                if (certificateUtil.isEmpty(userName)) {
+                    certificateUtil.addCertificate(userName, data);
+                    authenticationContext.setSubject(AuthenticatedUser
+                            .createLocalAuthenticatedUserFromSubjectIdentifier(userName));
+                } else {
+                    certificateUtil.validateCerts(userName, data);
```

Comment:
Here you are validating the certificate. But you are continuing the flow even it is not a valid certificate. Don't we have to do anything in case of invalid certificate. I think otherwise there is no point in doing the validation check here?